wagtail/query.py

```diff
@@ -253,7 +253,8 @@ def not_public(self):
         """
         This filters the QuerySet to only contain pages that are in a private section
         """
-        return self.exclude(self.public_q())
+        public_objects = self.public()
+        return self.exclude(id__in=[public_object.id for public_object in public_objects])
```

Comment:
Issue: Sadly this won't scale with a large number of pages.

Reply:
@RealOrangeOne can you suggest a solution?

Reply:
> @RealOrangeOne can you suggest a solution?

I'm not sure which component or test checks scaling with more pages. I guess it is ```test_not_public``` function in ```/wagtail/tests/test_page_queryset/TestPageQuerySet```, which is testing amount of query to db. All previous solution increase amount of query in this test. In your solution, you are using a generator which make an extra query. Because ```id__in``` can be ```QuerySet``` and combine in one SQL.  If I am right, the possible solution is
```
    def not_public(self):
        """
        This filters the QuerySet to only contain pages that are in a private section
        """
        public_objects = self.public().values_list('id', flat=True)
        return self.exclude(id__in=public_objects)
```
Please, let me know if that helped

Reply:
That definitely scales much better, but it's still not really ideal to have to do a separate query